### ISSUE TITLE
How to validate JSON null values for optional fields?

### ISSUE THREAD

**Issue by byrman** [tiangolo/sqlmodel#230]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from fastapi import FastAPI
from pydantic import BaseModel, validator
from sqlmodel import SQLModel


# NB: Inheriting from BaseModel!
class HeroUpdateA(BaseModel):
    name: Optional[str]
    secret_name: Optional[str]
    age: Optional[int]

    @validator("name", "secret_name", "age")
    def reject_none(cls, v):
        assert v is not None
        return v


# NB: Inheriting from SQLModel!
class HeroUpdateB(SQLModel):
    name: Optional[str]
    secret_name: Optional[str]
    age: Optional[int]

    @validator("name", "secret_name", "age")
    def reject_none(cls, v):
        assert v is not None
        return v


app = FastAPI()


# A PATCH request with payload {"age": 25} succeeds.
# name and secret_name are NOT validated.
@app.patch("/heroes-a/{id}")
async def patch_hero_a(obj: HeroUpdateA):
    pass


# A PATCH request with payload {"age": 25} fails.
# name and secret_name are validated.
@app.patch("/heroes-b/{id}")
async def patch_hero_b(obj: HeroUpdateB):
    pass
```

### Description

Suppose we have a `Hero(SQLModel, table=True)` model with no optional fields (so `name`, `secret_name`, and `age` are not nullable in the underlying database). To properly support PATCH, a HeroUpdate model is needed with all fields optional, because we don't know in advance what field(s) a user is going to supply.

Since the fields are not nullable, we want a validator to reject _explicit_ JSON null values, e.g. a payload like this: `{"name": null, "age": 25}`, otherwise our user might hit an Internal Server Error (HTTP 500) upon saving a hero to the database. These links propose a clever solution to this problem by writing a validator:

- https://github.com/samuelcolvin/pydantic/issues/1223
- https://github.com/tiangolo/fastapi/issues/3776

This solution works, because, _by default_, validators aren't called on arguments that are not provided. So a payload like `{"age": 25}` is fine and will not trigger validation of `name` and `secret_name`, even though these fields have `None` as a default.

To my surprise, however, the proposed solution does not work when inheriting from SQLModel, only when inheriting from BaseModel. In case of SQLModel, validation always fires, even when a field is not explicitly provided. Since SQLModel is powered by Pydantic, I expected similar behaviour.

If not a bug, what would be an elegant way to accomplish validation like this using SQLModel?

### Operating System

Linux

### Operating System Details

I'm using this Docker image: tiangolo/uvicorn-gunicorn-fastapi:python3.8-slim

### SQLModel Version

0.0.6

### Python Version

3.8.12

**Comment by byrman** [comment]
OK, for fields of type `str`, this could be solved by not making them `Optional`, but assign the empty string as a default instead: `name : str = ""` and skipping them while updating using the `exclude_unset` idiom. For many other types, however, this approach is cumbersome, because there is no sensible default (e.g. `uuid`, `AnyHttpUrl`, etc. Assigning fake defaults is an ugly solution in that case.

**Comment by byrman** [comment]
The problem appears to be in sqlmodel/main.py:
```
    # From Pydantic, override to enforce validation with dict
    @classmethod
    def validate(cls: Type["SQLModel"], value: Any) -> "SQLModel":
        if isinstance(value, cls):
            return value.copy() if cls.__config__.copy_on_model_validation else value

        value = cls._enforce_dict_if_root(value)
        if isinstance(value, dict):
            values, fields_set, validation_error = validate_model(cls, value)
            if validation_error:
                raise validation_error
            model = cls(**values)
            # Reset fields set, this would have been done in Pydantic in __init__
            object.__setattr__(model, "__fields_set__", fields_set)
            return model
        elif cls.__config__.orm_mode:
            return cls.from_orm(value)
        elif cls.__custom_root_type__:
            return cls.parse_obj(value)
        else:
            try:
                value_as_dict = dict(value)
            except (TypeError, ValueError) as e:
                raise DictError() from e
            return cls(**value_as_dict)
```
I will annotate what happens when patching a hero's age:
```
            # value = {"age": 25}

            values, fields_set, validation_error = validate_model(cls, value)

            # values = {'name': None, 'secret_name': None, 'age': 25}

            # There is not validation_error at this point, because, by default, pydantic
            # validators are not called on arguments that are not explicitly provided.

            if validation_error:
                raise validation_error

           # Now the situation if totally different: here, we are explicitly
           # providing arguments that won't pass validation!

           model = cls(**values)

```
I have made a pull request (including a test) that fixes this issue: https://github.com/tiangolo/sqlmodel/pull/253

**Comment by tiangolo** [comment]
Thanks for all the explanation and the fix @byrman! :bow: 

For completeness: this was fixed by @byrman in https://github.com/tiangolo/sqlmodel/pull/253. The fix will be available in SQLModel `0.0.7`, released in the next hours. :tada:
